### ISSUE TITLE
Secondary root hash in Etch for embedded peers

### ISSUE THREAD

**Issue by helins** [Convex-Dev/convex#346]
I was wondering how sensible it would be adding an extra secondary root. This would allow an application embedding a peer to share a same Etch instance with the peer itself. I recall @pedrorgirardi expressed the desire of using Etch instead of Datalevin for `convex.world`, this would make that kind of use cases more efficient.

**Comment by helins** [comment]
Yes but for instance the current peer implementation writes directly to the root and there is nothing you can do about it. Either this bit should be more flexible (providing a mapping function?) or there exist the possibility of having more than 1 root so that parties which do not need to coordinate effectively don't (peer don't care about application state and vice-versa).

**Comment by mikera** [comment]
Well you could have a map of `{app-id1 app-data1 ....}` stored as the root. It's really up to the Peer implementation to decide what format is used, this is outside the protocol spec.

We could make this pluggable I guess, have a default implementation with the current behaviour but you can override it if you like?

**Comment by helins** [comment]
Indeed peers can take many liberties but the default implementation should be as flexible as possible. Now one will currently or in the envisioned future build a peer from scratch :)

**Comment by mikera** [comment]
OK well I'm pretty happy with the proposal to let people customise the root hash persistence if they wish. Also remember there's nothing to stop someone storing their own root hash elsewhere, they don't need to use the built-in root hash facility.

**Comment by helins** [comment]
It would be helpful prioritizing this for 2 reasons:

- `convex-web` uses Datalevin while it looks like using sharing an Etch instance with the peer server would be enough. Datalevin uses illegal reflection so we cannot upgrade to Java 15 && it uses some native bindings which forces me to emulate a JVM in x86 making dev quite slow.

- In the future, Convex Shell will need to share an Etch instance in case of running a peer server.

---

The ideal solution would allow specifying a number of roots when creating an instance (defaulting to `1`). When creating a peer server, one could specify a "root index" to use (defaulting to `0`).

Given a discussion we had some time ago, if I'm not mistaken, this would allow any concurrent use of Etch (in-process) without having to synchronize anything at the root level (each consumer writes to its own root).

**Comment by mikera** [comment]
I'm working on the assumption that you probably shouldn't set the root too often, so locking and map performance hit is probably OK?

**Comment by helins** [comment]
Is it because it would mess the structure of Etch too much having more than 1 root?

Currently, peers and the Shell probably don't need to write too often. However, it would be more future-proof not assuming that. Especially if we diversify to Orbit-like, torrent-like ideas and whatnot.

**Comment by mikera** [comment]
I'm not concerned so much about the structure of Etch but having a sane API. Some considerations:

- If you have numbered slots, how do you allocate these? If you open an arbitrary Etch database, you wouldn't have much of a clue what they are. 
- How might you reconfigure, add new peers to a running instance etc?

A fixed number of slots looks like it could cause trouble, so a map is probably a better fit?

**Comment by helins** [comment]
Adding/removing "slots" over time is a solid concern. You have convinced me 👍

**Comment by brittleboye** [comment]
Closing as superseded by the lattice root design. An application embedding a peer shares the store through its own region under the root `KeyedLattice` (convex-core/docs/LATTICE_REGIONS.md, for example `:local` or an application keyword) rather than through a second Etch root. The thread already concluded that a fixed slot scheme was undesirable.
